File: frontend/src/types/simulation.ts
/**
 * Simulation-related types.
 */

import type { DeviceType, Timestamp, UUID, NumericArray } from './common';

/** Simulation backend types */
export type SimulationBackend = 'mujoco' | 'isaac' | 'pybullet' | 'dummy';

/** Available simulation tasks */
export type SimulationTask =
  | 'reach'
  | 'push'
  | 'pick_and_place'
  | 'insert'
  | 'custom';

/** Simulation session status */
export type SimulationStatus =
  | 'created'
  | 'running'
  | 'paused'
  | 'completed'
  | 'failed';

/** Configuration for creating a simulation session */
export interface SimulationConfig {
  task: SimulationTask;
  backend?: SimulationBackend;
  num_envs?: number;
  device?: DeviceType;
  seed?: number;
  render?: boolean;
  max_episode_steps?: number;
}

/** Simulation session response */
export interface SimulationSession {
  id: UUID;
  task: SimulationTask;
  backend: SimulationBackend;
  status: SimulationStatus;
  current_step: number;
  created_at?: Timestamp;
  metadata?: Record<string, unknown>;
}

/** Observation from simulation environment */
export interface Observation {
  joint_pos: NumericArray;
  joint_vel: NumericArray;
  ee_pos: [number, number, number];
  ee_quat: [number, number, number, number];
  goal_pos?: [number, number, number];
  object_pos?: [number, number, number];
  object_quat?: [number, number, number, number];
  gripper_state?: number;
  force_torque?: NumericArray;
  rgb?: string; // Base64 encoded image
  depth?: string; // Base64 encoded depth map
}

/** Full simulation state */
export interface SimulationState {
  time: number;
  qpos: NumericArray;
  qvel: NumericArray;
  qacc?: NumericArray;
  ctrl?: NumericArray;
  xpos?: Record<string, [number, number, number]>;
  xquat?: Record<string, [number, number, number, number]>;
  contact_forces?: NumericArray;
  sensor_data?: Record<string, number | NumericArray>;
}

/** Request to step the simulation */
export interface StepRequest {
  action: NumericArray;
  render?: boolean;
}

/** Result of stepping the simulation */
export interface StepResponse {
  observation: Observation;
  reward: number;
  terminated: boolean;
  truncated: boolean;
  info: StepInfo;
}

/** Additional info from simulation step */
export interface StepInfo {
  success?: boolean;
  is_success?: boolean;
  timeout?: boolean;
  collision?: boolean;
  distance_to_goal?: number;
  episode_length?: number;
  episode_reward?: number;
  [key: string]: unknown;
}

/** Request to reset the simulation */
export interface ResetRequest {
  seed?: number;
  options?: Record<string, unknown>;
}

/** Observation response wrapper */
export interface ObservationResponse {
  joint_pos: NumericArray;
  joint_vel: NumericArray;
  ee_pos: [number, number, number];
  ee_quat: [number, number, number, number];
  goal_pos?: [number, number, number];
}

/** State response wrapper */
export interface StateResponse {
  time: number;
  qpos: NumericArray;
  qvel: NumericArray;
}

/** Render configuration */
export interface RenderConfig {
  width?: number;
  height?: number;
  camera_id?: number | string;
  mode?: 'rgb' | 'depth' | 'segmentation';
}

/** Rendered frame */
export interface RenderedFrame {
  data: string; // Base64 encoded
  width: number;
  height: number;
  format: 'rgb' | 'rgba' | 'depth';
  timestamp: number;
}

/** Environment info */
export interface EnvironmentInfo {
  name: string;
  observation_space: SpaceInfo;
  action_space: SpaceInfo;
  reward_range: [number, number];
  max_episode_steps?: number;
}

/** Space specification */
export interface SpaceInfo {
  type: 'box' | 'discrete' | 'multi_discrete' | 'dict';
  shape?: number[];
  dtype?: string;
  low?: NumericArray;
  high?: NumericArray;
  n?: number;
  nvec?: number[];
  spaces?: Record<string, SpaceInfo>;
}

/** Batch step request for vectorized environments */
export interface BatchStepRequest {
  actions: NumericArray[]; // [num_envs, action_dim]
}

/** Batch step response */
export interface BatchStepResponse {
  observations: Observation[];
  rewards: number[];
  terminated: boolean[];
  truncated: boolean[];
  infos: StepInfo[];
}

File: src/arcs/api/schemas/simulation.py
"""Simulation-related schemas."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class SimulationConfig(BaseModel):
    """Configuration for creating a simulation session."""

    task: Literal["Reach", "Pick", "Place", "Push"] = Field(
        default="Reach",
        description="Simulation task type"
    )
    backend: Literal["dummy", "isaacgym", "mujoco", "pybullet"] = Field(
        default="dummy",
        description="Simulation backend"
    )
    num_envs: int = Field(
        default=1,
        ge=1,
        le=4096,
        description="Number of parallel environments"
    )
    device: Literal["cpu", "cuda"] = Field(
        default="cpu",
        description="Compute device"
    )


class SimulationSessionResponse(BaseModel):
    """Simulation session response."""

    id: str = Field(..., description="Session ID")
    task: str = Field(..., description="Task type")
    backend: str = Field(..., description="Backend type")
    status: Literal["idle", "running", "paused", "completed", "error"] = Field(
        default="idle",
        description="Session status"
    )
    current_step: int = Field(default=0, description="Current simulation step")
    created_at: datetime = Field(default_factory=datetime.utcnow)


class ObservationResponse(BaseModel):
    """Observation from the simulation."""

    proprio: list[float] = Field(..., description="Proprioceptive state (joint angles, velocities)")
    rgb: list[list[list[float]]] | None = Field(default=None, description="RGB image [H, W, 3]")
    depth: list[list[float]] | None = Field(default=None, description="Depth image [H, W]")
    tactile: list[float] | None = Field(default=None, description="Tactile sensor readings")
    timestep: int = Field(..., description="Simulation timestep")
    metadata: dict[str, Any] = Field(default_factory=dict)


class StateResponse(BaseModel):
    """Simulation state response."""

    joint_pos: list[float] = Field(..., description="Joint positions (7 DOF)")
    joint_vel: list[float] = Field(..., description="Joint velocities (7 DOF)")
    object_states: dict[str, list[float]] = Field(
        default_factory=dict,
        description="Object states"
    )
    timestamp: float = Field(..., description="State timestamp")


class ResetRequest(BaseModel):
    """Request to reset simulation."""

    seed: int | None = Field(default=None, description="Random seed for reset")


class StepRequest(BaseModel):
    """Request to step simulation."""

    action: list[float] = Field(
        ...,
        min_length=7,
        max_length=7,
        description="Action vector (7 DOF joint commands)"
    )


class StepResponse(BaseModel):
    """Response from simulation step."""

    observation: ObservationResponse
    reward: float = Field(..., description="Step reward")
    terminated: bool = Field(..., description="Episode terminated")
    truncated: bool = Field(..., description="Episode truncated")
    info: dict[str, Any] = Field(default_factory=dict, description="Additional info")
